plugins/maciVoting/contexts/CoordinatorContext.tsx

```diff
@@ -191,13 +179,64 @@ export const CoordinatorProvider = ({ children }: { children: ReactNode }) => {
     [getAuthorizationHeader, getPublicKey, publicClient]
   );
 
+  const checkMergeStatus = async (pollId: number) => {
+    const { address: pollAddress } = await getPoll({
+      maciAddress: PUBLIC_MACI_ADDRESS,
+      pollId,
+      signer,
+    });
+    const pollAbi = parseAbi(["function stateMerged() view returns (bool)"]);
+    const statemMerged = await publicClient.readContract({
+      address: pollAddress as `0x${string}`,
+      abi: pollAbi,
+      functionName: "stateMerged",
+    });
+    return statemMerged;
+  };
+
+  const finalizeProposal = useCallback(async (pollId: number) => {
+    setFinalizeStatus("merging");
+    const hasMerged = await checkMergeStatus(pollId);
+    if (!hasMerged) {
+      const mergeResult = await merge(pollId);
+      if (!mergeResult.success) {
+        return;
+      }
+    }
+    setFinalizeStatus("merged");
+
+    // TODO: get these values
+    const encryptedCoordinatorPrivateKey = "";
+    const startBlock = 0;
```

Comment:
The start block should be PUBLIC_MACI_DEPLOYMENT_BLOCK or the poll deployment block (maybe fetching the poll block would require more computation) and the end block would be the block when the poll finalizes (we could use `getFutureBlockNumberAtTimestamp` https://github.com/privacy-scaling-explorations/maci-aragon-osx-gov-app/blob/b81680da0567639f27acf1ec22121d9c6d78109c/plugins/maciVoting/utils/blockAtTimestamp.ts#L52 or similar)

Reply:
Does `getFutureBlockNumberAtTimestamp` handle cases where the future timestamp is in the past? The poll should have ended by now or not?

Reply:
`getFutureBlockNumberAtTimestamp` does not account for timestamps in the past (it calculates the difference between the given timestamp and the current timestamp. If it is in the past, the result is negative and breaks the logic). We could implement some logic to avoid this.

In the same file there is `getBlockNumberAtTimestamp` which according to the Aragon team: 

*// This results in a crap ton of requests to the RPC node, so we need a better way.*



In the meantime, we will not be needing `getFutureBlockNumberAtTimestamp` (checkout changes in [feat/better-ui branch](https://github.com/privacy-scaling-explorations/maci-aragon-osx-gov-app/pull/22)). I created a issue (https://github.com/privacy-scaling-explorations/maci-aragon-osx-gov-app/issues/21) to address this later on.